<a href="https://colab.research.google.com/github/TAEO2474/python-dev/blob/main/505_%EA%B8%B0%EA%B3%84%ED%95%99%EC%8A%B5_Model_Selection_%ED%95%98%EC%9D%B4%ED%8D%BC%ED%8C%8C%EB%9D%BC%EB%AF%B8%ED%84%B0_%ED%8A%9C%EB%8B%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 최종 목표
- 어떤 꽃이 들어오면,

- 그 꽃이 "세토사(setosa)", "버시컬러(versicolor)", "버지니카(virginica)"
- 3개 중에 어떤 종류인지 맞히는 똑똑한 프로그램(모델)을 만드는 것!


### 🌼 <예시>

- 우리는 지금 **꽃을 구별하는 AI 선생님**을 만들고 있어요.

- 그런데 이 선생님은 **눈으로 꽃을 보지 못해요.**

- 대신 숫자로 알려줘야 해요. 이렇게요:

| 꽃 이름 | 꽃잎 길이 | 꽃잎 너비 | 꽃받침 길이 | 꽃받침 너비 |
| --- | --- | --- | --- | --- |
| setosa | 1.4 | 0.2 | 5.1 | 3.5 |
| versicolor | 4.5 | 1.5 | 6.0 | 2.5 |
| virginica | 5.6 | 2.0 | 6.5 | 3.0 |

 -그래서 우리는 150송이의 꽃을 보여주면서,

- *“이 꽃은 setosa야, 이건 virginica야…”** 하면서 알려줘요.

# ✅ 그런데 우리는 왜 6송이를 썼을까?


- 이유는 처음엔 K=1부터 K=19까지 여러 값을 실험해봤어요.

- K=6일 때 평균 정확도가 제일 높았어요.( 즉 최고의 파라미터가 K(이웃수)=6)

- 그래서 기본값 5보다 더 좋은 성능을 내는 K=6을 선택한 거예요!

# <앞으로 진행과정>
| 단계  | 한 일                   | 쉽게 말하면                       |
| --- | --------------------- | ---------------------------- |
| 1단계 | 데이터를 준비함              | 꽃들의 길이, 너비를 숫자로 정리함          |
| 2단계 | 여러 개의 K값으로 실험함( ✅ 하이퍼파라미터 튜닝)        | “이웃 몇 송이로 비교하면 잘 맞출까?” 실험함 (✅ 교차검증으로 성능 평가)  |
| 3단계 | K=6이 제일 성능 좋음 ( ✅ 하이퍼파라미터 튜닝)            | “6송이 보면 가장 정확하게 맞추더라!”       |
| 4단계 | 최종 모델 만들기             | "좋아! 6송이로 비교하는 AI 모델 만들자!"  |
| 5단계 | 모든 데이터를 학습시킴          | 제작한 모델 한테 모든 꽃을 보여주며 외우게 함       |
| 6단계 | 이제 새로운 꽃이 오면 맞출 수 있음! | 진짜 시험을 보는 거야! 꽃을 주면 AI가 맞춘다!

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#iris-dataset" data-toc-modified-id="iris-dataset-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>iris dataset</a></span></li><li><span><a href="#score" data-toc-modified-id="score-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>score</a></span></li><li><span><a href="#Hyper-Parameter" data-toc-modified-id="Hyper-Parameter-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Hyper Parameter</a></span><ul class="toc-item"><li><span><a href="#3-way-holdout" data-toc-modified-id="3-way-holdout-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>3-way holdout</a></span></li><li><span><a href="#GridSearchCV" data-toc-modified-id="GridSearchCV-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>GridSearchCV</a></span></li></ul></li><li><span><a href="#wine-dataset" data-toc-modified-id="wine-dataset-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>wine dataset</a></span><ul class="toc-item"><li><span><a href="#GridSearchCV-2" data-toc-modified-id="GridSearchCV-2-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>GridSearchCV 2</a></span></li></ul></li><li><span><a href="#GridSearchCV-+-Pipeline" data-toc-modified-id="GridSearchCV-+-Pipeline-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>GridSearchCV + Pipeline</a></span><ul class="toc-item"><li><span><a href="#GridSearchCV와-Pipeline을-연동하기" data-toc-modified-id="GridSearchCV와-Pipeline을-연동하기-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>GridSearchCV와 Pipeline을 연동하기</a></span></li><li><span><a href="#GridSearchCV로-최적의-알고리즘-찾기" data-toc-modified-id="GridSearchCV로-최적의-알고리즘-찾기-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>GridSearchCV로 최적의 알고리즘 찾기</a></span></li><li><span><a href="#알고리즘마다-하이퍼-파라미터-찾기" data-toc-modified-id="알고리즘마다-하이퍼-파라미터-찾기-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>알고리즘마다 하이퍼 파라미터 찾기</a></span></li></ul></li></ul></div>

## iris dataset (아이리스 꽃 데이터셋)

In [ ]:
from sklearn.datasets import load_iris

In [ ]:
data = load_iris()
print(dir(data))

['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [ ]:
import pandas as pd

In [ ]:
data_pd = pd.DataFrame(data.data, columns=data.feature_names)
target_pd = pd.DataFrame(data.target, columns=['target'])
iris = pd.concat([data_pd, target_pd], axis=1)

In [ ]:
iris

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0         0  
1         0  
2         0  
3         0  
4         0  
..      ...  
145       2  
146       2  
147       2  
148       2  
149       2  

[150 rows x 5 columns]

In [ ]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


## ✅ score란?
- 모델이 얼마나 잘 맞추는지를 숫자로 보여주는 함수예요.
- 보통은 **정확도(accuracy)**를 나타냅니다.



- 대충의 성능 파악(dummy)

- 실제 모델은 내가 가진 모든 데이터로 만든다.

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
# 2way hold out
# iris.iloc[:, :-1]는 iris 데이터프레임에서 마지막 열을 제외한 모든 데이터를 선택합니다.
# iris.target은 iris 데이터프레임에서 target 열만 선택합니다.
X_train, X_test, y_train, y_test = train_test_split(iris.iloc[:, :-1], iris.target)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
from sklearn.neighbors  import KNeighborsClassifier

In [ ]:
# KNeighborsClassifier 모델을 생성합니다.
knn = KNeighborsClassifier()

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [ ]:
# 행은 실제, 열은 예측한 값 이다.
# 실제 0인데, 0으로 예측한 값 13개
# 실제 1인데, 1로 예측한 값 10개 & 2로 예측한 값 0개
# 실제 2인데, 1로 예측한 값 2개 & 2로 예측한 값 13개
pred=knn.predict(X_test)
confusion_matrix(y_test,pred )
#confusion_matrix(y_test, knn.predict(X_test))

array([[ 8,  0,  0],
       [ 0, 14,  0],
       [ 0,  0, 16]])

In [ ]:
# 0, 1, 2는 다음과 같다.
data.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

위의 confusion matirx는 다음과 같다.

|   |setosa로 예측|versicolor로 예측|virginica로 예측|
|---|---|---|---|
|실제 setosa|13|0|0|
|실제 versicolor|0|10|0|
|실제 virginica|0|2|13|

---

이진 분류하는 confusion matrix에서 각 위치를 다음과 같이 이름 붙인다.

|||Predict||
|---|---|---|---|
|   |    |True|False|
|Actual|True|True Positive|False Negative|
|   |False|False Positive|True Negative|

- False Positivie는 1종 오류  
- False Negative는 2종 오류

1종 오류와 2종 오류 중에서 어떤 것이 더 중요한지는, 의사결정자의 판단에 따라 다르다.

---

Confusion matrix에 기반하여 다음과 같은 모델의 성능 지표가 있다.  
정확도는 맹점이 있기 때문에 다른 평가지표도 함께 살펴볼 필요가 있다.

- accuracy(정확도)  
    전체 데이터 중에 예측이 맞은 비율  
>   $accuracy = \frac{TP+TN}{TP+FP+TN+FN}$

- precision(정밀도)  
    양성으로 예측한 것 중에서 양성이 맞는 비율  
    즉, 예측한 거 중에 맞은 비율  
>   precision = \frac{TP}{TP+FP}$

- recall(재현률)  
    실제 양성인 것  중에서 양성으로 예측한 비율  
    즉, 맞은 비율  
>  $reacll = \frac{TP}{TP+FN}$

multi class 일 경우는, 각 클래스 별로 precision을 구하고 평균을 내면 최종 precision이 된다.  
최종 recall도 마찬가지 방법으로 구한다.

In [ ]:
from sklearn.model_selection import cross_val_score

In [ ]:
from sklearn.metrics import recall_score,precision_score

In [ ]:
# accuracy 대신 recall로 cross validation score를 측정할 수 있다.
cross_val_score(KNeighborsClassifier(), X_train, y_train, scoring='recall_micro', cv=5)

array([0.95652174, 0.91304348, 0.95454545, 1.        , 0.95454545])

In [ ]:
cross_val_score(KNeighborsClassifier(), X_train, y_train, scoring='precision_micro')

array([0.95652174, 0.91304348, 0.95454545, 1.        , 0.95454545])

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
# 내부적으로 confusion matrix를 만든 후 평가 지표별로 점수를 계산해준다.
#print(classification_report(y_test, knn.predict(X_test)))
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         8
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00        16

    accuracy                           1.00        38
   macro avg       1.00      1.00      1.00        38
weighted avg       1.00      1.00      1.00        38



## Hyper Parameter

- 하이퍼 파라미터에 따라서 성능이 달라진다.  
고로, 최적의 하이퍼 파라미터를 찾는 것이 중요하다.
- 즉, 인간이 하이퍼파라미터(매개변수)를  넣어주고 컴퓨터가 최고의 파라미터(매개변수)를 찾는 것이다

In [ ]:
# 단순하게 하이퍼 파라미터를 for문으로 찾는 방법을 생각할 수 있다.
temp = []
for i in range(1, 20):  # 몇 개의 이웃을 보고 분류할 것인가? 1~19개로 설정 -> n_neighbors=1부터 n_neighbors=19까지 시도해보겠다는 뜻
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train, y_train)
    temp.append(knn.score(X_test, y_test))

# **테스트용 19개 아이리스 꽃들의 특징(X_test)**를 보고,
# 각각의 k값으로 라벨(꽃의 종류)을 예측해서,
# y_test와 비교했을 때 맞춘 비율 = 정확도가 리스트에 저장된 값들입니다.

In [ ]:
temp

[0.9736842105263158,
 0.8947368421052632,
 0.9736842105263158,
 0.9736842105263158,
 1.0,
 0.9736842105263158,
 1.0,
 0.9473684210526315,
 0.9736842105263158,
 0.9473684210526315,
 0.9736842105263158,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9736842105263158,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9736842105263158]

In [ ]:
import numpy as np

In [ ]:
# argmax는 가장 큰 값의 index를 알려준다.
# n_neighbors가 1일때 성능이 가장 좋다.
np.argmax(temp)

np.int64(4)

### 교차검증(3way holdout)

데이터를 총 3등분한다.
- train set : 모델 학습에 사용한다.
- validationt set : 하이퍼 파라미터를 찾기 위해서 사용한다.
- test set : 한번도 안 본 데이터로 모델의 성능을 측정하기 위해서 사용한다.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(iris.iloc[:, :-1], iris.target)

In [ ]:
X_train, X_validation, y_train, y_validation = train_test_split(X_train, y_train)

 **"하이퍼파라미터 튜닝"**을 위한 3-way holdout 방식을 사용해서 최적의 K 값을 찾는 과정

K=1

In [ ]:
# 하이퍼 파라미터를 찾기 위해서 3-way holdout을 한다.
temp = []
for i in range(1, 20):
    knn = KNeighborsClassifier(n_neighbors=i) # 이웃의 수K를 보고 검사를 한다 K=1 (이웃1개만 보고 검사한다)
    knn.fit(X_train, y_train)
    temp.append(knn.score(X_validation, y_validation))

In [ ]:
temp

[0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9736842105263158,
 0.9736842105263158,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9736842105263158,
 0.9736842105263158,
 0.9736842105263158,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315,
 0.9473684210526315]

### <결과해석>
- temp[0] = 0.9473684210526315 (k=1 이웃수를 1개로 잡았을때 설정한 KNN 모델이 **검증 데이터(validation set)**에서
**정확도 94.7%**를 기록했다는 뜻

In [ ]:
# 3way holdout으로 했을 때, 하이퍼 파라미터 n_neighbors가 6일때 가장 성능이 좋다.
# temp[5] = K=6일 때 정확도 = **0.973 (최고)**
np.argmax(temp) # temp의 가장 성능이 좋은 인데스가 출력됨

np.int64(5)

#### cross_val_score
- 사이킷런(Scikit_Learn)에서는 교차검증(K-Fold, StratifiedKFold)을 쉽게 할수 있도록 제공하는 API 이다.
- scikit-learn 0.22 버전부터 기본적으로 5-겹 교차 검증으로 바뀌었다.
- cross_val_score에서 사용되는 파라미터
  - estimator : 분류(Classifier) 또는 회귀(Regressor)인지 구분
  - X : feature  데이터 세트
  - y : class 데이터 세트
  - scoring : 예측 성능 평가 지표('accuracy', 'neg_brier_score', 'top_k_accuracy'...)
  - cv : 교차 검증 필드 수
- y가바이너리 또는 다중 클래스, StratifiedKFold가 사용되며, 전체적으로 그 외의 경우 KFold 사용된다.    
- 수행 후 반환값은  scoring 파라미터로 지정된 측정값을 배열 형태로 반환한다.


In [ ]:
# cross_val_score로 하이퍼 파라미터 찾기
# 이 방법은 모든 데이터를 사용하기 때문에 위의 temp 방식 보다 많이 사용한다.
temp2 = [] # K값에 따른 평균 정확도를 저장

for i in range(1, 20):
    temp2.append(
        cross_val_score(
            KNeighborsClassifier(n_neighbors=i),
            iris.iloc[:, :-1],     # 특징 (X)
            iris.target,           # 레이블 (y)
            cv=5                   # 5겹 교차 검증
        ).mean()                   # 5개의 점수 평균을 저장
    )


#  교차 검증 분할 수 (`cv=5`이면 5-fold cross validation)
# 기본값은 `cv=5` (scikit-learn 0.22부터)
# `StratifiedKFold`은 분류 문제에서 클래스 비율을 **고르게 유지**

In [ ]:
# temp와 최적의 하이퍼 파라미터가 다르게 나왔다.
# temp[5] = K=6일 때 정확도 = **0.973 (최고)**
np.argmax(temp2)

np.int64(5)

In [ ]:
temp2

[np.float64(0.96),
 np.float64(0.9466666666666665),
 np.float64(0.9666666666666668),
 np.float64(0.9733333333333334),
 np.float64(0.9733333333333334),
 np.float64(0.9800000000000001),
 np.float64(0.9800000000000001),
 np.float64(0.9666666666666668),
 np.float64(0.9733333333333334),
 np.float64(0.9800000000000001),
 np.float64(0.9800000000000001),
 np.float64(0.9800000000000001),
 np.float64(0.9733333333333334),
 np.float64(0.9666666666666666),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666666),
 np.float64(0.9666666666666668)]

<출력결과>
```
[np.float64(0.96),   # temp2[0]
 np.float64(0.9466666666666665),   # temp2[1]  
 np.float64(0.9666666666666668),   # temp2[2]  
 np.float64(0.9733333333333334),  # temp2[3]  
 np.float64(0.9733333333333334),  # temp2[4]  
 np.float64(0.9800000000000001),  # temp2[5] K=6  -> 바로 여기가 최고의 파라미터발견 -> 최고의 정확도

 np.float64(0.9800000000000001), # temp2[6]  -> 바로 여기가 최고의 파라미터발견 -> 최고의 정확도
 np.float64(0.9666666666666668),
 np.float64(0.9733333333333334),
 np.float64(0.9800000000000001),
 np.float64(0.9800000000000001),
 np.float64(0.9800000000000001),
 np.float64(0.9733333333333334),
 np.float64(0.9666666666666666),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666668),
 np.float64(0.9666666666666666),
 np.float64(0.9666666666666668)]

K = 6 → 정확도 = 0.9800000000000001
K = 7 → 정확도 = 0.9800000000000001
K = 10 → 정확도 = 0.9800000000000001
K = 11 → 정확도 = 0.9800000000000001
K = 12 → 정확도 = 0.9800000000000001
 ```



# **최적의 하이퍼파라미터(K=6)**를 사용해서 최종 KNN 모델을 전체 데이터를 이용해 학습하는 과정
- 아래  코드는 KNN 알고리즘을 사용해서
아이리스 꽃의 종류를 분류(Classification)하기 위한 최종 모델을 학습하는 과정입니다.

In [ ]:
# 찾아낸 하이퍼 파라미터로 최종 모델을 만든다.
# temp[5] = K=6일 때 정확도 = **0.973 (최고)**
# 최종 모델은 모든 데이터를 써서 학습한다.
knn_final = KNeighborsClassifier(n_neighbors=6)
knn_final.fit(iris.iloc[:, :-1], iris.target)

KNeighborsClassifier(n_neighbors=6)

#### cross_validate
- cross_val_score와 비슷하지만, 분할마다 훈련과 테스트에 걸린 시간을 담은 딕셔너리를 반환한다
 - fit_time : 훈련 시간
 - score_time : 테스트 시간
 - train_score : 훈련 점수
 - test_score: 테스트 점수

In [ ]:
from sklearn.model_selection import cross_validate
import numpy as np


temp3 = []
for i in range(1, 20):
    temp3.append(cross_validate(KNeighborsClassifier(n_neighbors=i),iris.iloc[:,:-1], iris.target, cv=5, return_train_score=True))

In [ ]:
temp3

[{'fit_time': array([0.0055294 , 0.00328565, 0.00558686, 0.00543141, 0.00320196]),
  'score_time': array([0.00691748, 0.00433922, 0.00469422, 0.00677991, 0.00431871]),
  'test_score': array([0.96666667, 0.96666667, 0.93333333, 0.93333333, 1.        ]),
  'train_score': array([1., 1., 1., 1., 1.])},
 {'fit_time': array([0.00297308, 0.00375533, 0.00394583, 0.00310445, 0.00447893]),
  'score_time': array([0.00431347, 0.00689387, 0.00435424, 0.00421834, 0.00415516]),
  'test_score': array([0.96666667, 0.93333333, 0.93333333, 0.9       , 1.        ]),
  'train_score': array([0.975     , 0.98333333, 0.975     , 0.98333333, 0.975     ])},
 {'fit_time': array([0.00347424, 0.00477934, 0.00302768, 0.00666356, 0.00309658]),
  'score_time': array([0.00459409, 0.00993299, 0.00414515, 0.00530934, 0.00523233]),
  'test_score': array([0.96666667, 0.96666667, 0.93333333, 0.96666667, 1.        ]),
  'train_score': array([0.95833333, 0.95833333, 0.96666667, 0.96666667, 0.95      ])},
 {'fit_time': array(

###✅ GridSearchCV
- 여러 하이퍼파라미터 조합을 자동으로 실험하고,
- 교차검증을 통해 가장 좋은 조합을 찾아주는 도구예요.

쉽게 말해:

"이웃 수를 1부터 20까지 다 시험해보고,
가장 성능 좋은 숫자를 자동으로 골라줘!"
이런 걸 해주는 똑똑한 도우미입니다 💡

- 교차검증(여러가지 기법들이 존재함)과 하이퍼 파라미터 튜닝(weight)값을 둘다 할 수 있는 API이다.

scikit-learn에서 하이퍼 파라미터를 편하게 찾는 기능을 제공하고 있다.

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
knn = KNeighborsClassifier()

In [ ]:
# scikit-learn의 모델들은 get_params를 공통적으로 가지고 있다.
# 알고리즘이 사용할 수 있는 하이퍼 파라미터를 알려준다.
knn.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [ ]:
# param_grid의 key는 get_params에 나오는 하이퍼 파라미터만 쓸 수 있다.
# param_grid의 value는 테스트해 볼 하이퍼 파라미터를 리스트로 넣는다.
grid = GridSearchCV(KNeighborsClassifier(), {'n_neighbors':[2,3,4,5,6,7]})

In [ ]:
grid.fit(iris.iloc[:, :-1], iris.target)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [2, 3, 4, 5, 6, 7]})

In [ ]:
# 가장 좋은 하이퍼 파라미터를 한번에 찾아준다.
grid.best_params_

{'n_neighbors': 6}

In [ ]:
# 가장 좋은 하이퍼 파라미터의 성능을 알려준다.
grid.best_score_

np.float64(0.9800000000000001)

In [ ]:
# grid search한 결과를 총 정리해준다.
grid.cv_results_

{'mean_fit_time': array([0.00336199, 0.00580478, 0.01332879, 0.00590229, 0.00813498,
        0.00602822]),
 'std_fit_time': array([0.00010228, 0.00303296, 0.00945245, 0.00220865, 0.00382023,
        0.00281052]),
 'mean_score_time': array([0.00457664, 0.00795999, 0.01007853, 0.01184049, 0.00774083,
        0.01086226]),
 'std_score_time': array([0.0003725 , 0.00323169, 0.00530567, 0.00720098, 0.00435098,
        0.00186895]),
 'param_n_neighbors': masked_array(data=[2, 3, 4, 5, 6, 7],
              mask=[False, False, False, False, False, False],
        fill_value=999999),
 'params': [{'n_neighbors': 2},
  {'n_neighbors': 3},
  {'n_neighbors': 4},
  {'n_neighbors': 5},
  {'n_neighbors': 6},
  {'n_neighbors': 7}],
 'split0_test_score': array([0.96666667, 0.96666667, 0.96666667, 0.96666667, 0.96666667,
        0.96666667]),
 'split1_test_score': array([0.93333333, 0.96666667, 0.96666667, 1.        , 1.        ,
        1.        ]),
 'split2_test_score': array([0.93333333, 0.93333333, 0

In [ ]:
# pandas로 만들면 보기 쉽다.
# column이 너무 많을때는 전치(T)해서 보면 편하다.
pd.DataFrame(grid.cv_results_).T

0                   1                   2  \
mean_fit_time                0.003362            0.005805            0.013329   
std_fit_time                 0.000102            0.003033            0.009452   
mean_score_time              0.004577             0.00796            0.010079   
std_score_time               0.000373            0.003232            0.005306   
param_n_neighbors                   2                   3                   4   
params             {'n_neighbors': 2}  {'n_neighbors': 3}  {'n_neighbors': 4}   
split0_test_score            0.966667            0.966667            0.966667   
split1_test_score            0.933333            0.966667            0.966667   
split2_test_score            0.933333            0.933333            0.966667   
split3_test_score                 0.9            0.966667            0.966667   
split4_test_score                 1.0                 1.0                 1.0   
mean_test_score              0.946667            0.966667            0.973333   
std_test_score               0.033993            0.021082            0.013333   
rank_test_score                     6                   5                   3   

                                    3                   4                   5  
mean_fit_time                0.005902            0.008135            0.006028  
std_fit_time                 0.002209             0.00382            0.002811  
mean_score_time               0.01184            0.007741            0.010862  
std_score_time               0.007201            0.004351            0.001869  
param_n_neighbors                   5                   6                   7  
params             {'n_neighbors': 5}  {'n_neighbors': 6}  {'n_neighbors': 7}  
split0_test_score            0.966667            0.966667            0.966667  
split1_test_score                 1.0                 1.0                 1.0  
split2_test_score            0.933333            0.966667            0.966667  
split3_test_score            0.966667            0.966667            0.966667  
split4_test_score                 1.0                 1.0                 1.0  
mean_test_score              0.973333                0.98                0.98  
std_test_score               0.024944             0.01633             0.01633  
rank_test_score                     3                   1                   1

In [ ]:
# 하이퍼 파라미터를 여러개 한번에 찾을 수 있다.
grid = GridSearchCV(KNeighborsClassifier(), {'n_neighbors':[2,3,4,5,6,7],
                                            'weights':['distance', 'uniform']})

In [ ]:
grid.fit(iris.iloc[:, :-1], iris.target)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [2, 3, 4, 5, 6, 7],
                         'weights': ['distance', 'uniform']})

In [ ]:
# params를 보면 입력된 하이퍼 파라미터들의 가능한 조합을 모두 계산했다.
pd.DataFrame(grid.cv_results_).T

0   \
mean_fit_time                                       0.006641   
std_fit_time                                        0.002968   
mean_score_time                                     0.008023   
std_score_time                                      0.002947   
param_n_neighbors                                          2   
param_weights                                       distance   
params             {'n_neighbors': 2, 'weights': 'distance'}   
split0_test_score                                   0.966667   
split1_test_score                                   0.966667   
split2_test_score                                   0.933333   
split3_test_score                                   0.933333   
split4_test_score                                        1.0   
mean_test_score                                         0.96   
std_test_score                                      0.024944   
rank_test_score                                           11   

                                                         1   \
mean_fit_time                                      0.005872   
std_fit_time                                       0.002459   
mean_score_time                                    0.006954   
std_score_time                                     0.002431   
param_n_neighbors                                         2   
param_weights                                       uniform   
params             {'n_neighbors': 2, 'weights': 'uniform'}   
split0_test_score                                  0.966667   
split1_test_score                                  0.933333   
split2_test_score                                  0.933333   
split3_test_score                                       0.9   
split4_test_score                                       1.0   
mean_test_score                                    0.946667   
std_test_score                                     0.033993   
rank_test_score                                          12   

                                                          2   \
mean_fit_time                                       0.005746   
std_fit_time                                        0.003729   
mean_score_time                                     0.004154   
std_score_time                                       0.00135   
param_n_neighbors                                          3   
param_weights                                       distance   
params             {'n_neighbors': 3, 'weights': 'distance'}   
split0_test_score                                   0.966667   
split1_test_score                                   0.966667   
split2_test_score                                   0.933333   
split3_test_score                                   0.966667   
split4_test_score                                        1.0   
mean_test_score                                     0.966667   
std_test_score                                      0.021082   
rank_test_score                                            7   

                                                         3   \
mean_fit_time                                       0.00935   
std_fit_time                                       0.003945   
mean_score_time                                     0.00795   
std_score_time                                     0.003967   
param_n_neighbors                                         3   
param_weights                                       uniform   
params             {'n_neighbors': 3, 'weights': 'uniform'}   
split0_test_score                                  0.966667   
split1_test_score                                  0.966667   
split2_test_score                                  0.933333   
split3_test_score                                  0.966667   
split4_test_score                                       1.0   
mean_test_score                                    0.966667   
std_test_score                                     0.021082   
rank_test_score                                       

## wine dataset

In [ ]:
from sklearn.datasets import load_wine

In [ ]:
data = load_wine()

In [ ]:
data_pd = pd.DataFrame(data.data, columns=data.feature_names)
data_pd['target'] = pd.DataFrame(data.target, columns=['target'])

In [ ]:
wine = data_pd.copy()

In [ ]:
wine

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  target  
0                            3.92   1065.0       0  
1                            3.40   1050.0       0  
2                            3.17   1185.0       0  
3                            3.45   1480.0       0  
4                            2.93    735.0       0  
..                            ...      ...     ...  
173                          1.74    740.0       2  
174                          1.56    750.0       2  
175                          1.56    835.0       2  
176                          1.62    840.0       2  
177                          1.60    560.0       2  

[178 rows x 14 columns]

In [ ]:
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
 13  targe

In [ ]:
print(dir(wine))

['T', '_AXIS_LEN', '_AXIS_ORDERS', '_AXIS_TO_AXIS_NUMBER', '_HANDLED_TYPES', '__abs__', '__add__', '__and__', '__annotations__', '__array__', '__array_priority__', '__array_ufunc__', '__arrow_c_stream__', '__bool__', '__class__', '__contains__', '__copy__', '__dataframe__', '__dataframe_consortium_standard__', '__deepcopy__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__divmod__', '__doc__', '__eq__', '__finalize__', '__floordiv__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__iadd__', '__iand__', '__ifloordiv__', '__imod__', '__imul__', '__init__', '__init_subclass__', '__invert__', '__ior__', '__ipow__', '__isub__', '__iter__', '__itruediv__', '__ixor__', '__le__', '__len__', '__lt__', '__matmul__', '__mod__', '__module__', '__mul__', '__ne__', '__neg__', '__new__', '__nonzero__', '__or__', '__pandas_priority__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', 

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(wine.iloc[:, 0:-1], wine['target'], random_state=140)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)


(133, 13)
(133,)
(45, 13)
(45,)


In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
# 전문 기업들은 전처리만 하고 나면, 자동화된 코드를 모두 만들어두었다.
# 함수로 만들어두면 재사용이 가능하다.

def all_models(X_train, y_train,  cv=10):
    ###################################################
    # 1. Logistic Regression                          #
    ###################################################
    param_grid_lr = {
        'C': [0.01, 0.1, 1, 10],
        'solver': ['liblinear'],  # 작은 데이터셋에서 안정적
        'penalty': ['l1', 'l2']
    }
    grid_lr = GridSearchCV(LogisticRegression(),param_grid_lr, cv=cv)
    grid_lr.fit(X_train, y_train)
    lr_pd = pd.DataFrame(grid_lr.cv_results_)
    ###################################################
    # 2. KNeighbors                                   #
    ###################################################
    param_grid_knn = {
        'n_neighbors': [3, 5, 7],
        'weights': ['uniform', 'distance'],
        'metric': ['euclidean', 'manhattan']
    }
    grid_knn = GridSearchCV(KNeighborsClassifier(), param_grid_knn , cv=cv)
    grid_knn.fit(X_train, y_train)
    knn_pd = pd.DataFrame(grid_knn.cv_results_)

    return pd.concat([lr_pd, knn_pd], axis=1)

# LogisticRegression().get_params()
-  LogisticRegression 모델의 모든 하이퍼파라미터와 그 기본값을 확인하는 코드입니다.

- **`LogisticRegression()`**: 로지스틱 회귀 모델의 인스턴스를 생성합니다.
- **`.get_params()`**: 이 메서드는 모델이 가지고 있는 모든 파라미터(하이퍼파라미터 포함)를 딕셔너리 형태로 반환합니다.

이 코드를 실행하면 다음과 같은 정보를 얻을 수 있습니다:

- **어떤 하이퍼파라미터를 조절할 수 있는지**: 모델의 성능을 튜닝하기 위해 변경할 수 있는 모든 하이퍼파라미터의 목록을 보여줍니다.
- **각 하이퍼파라미터의 기본값은 무엇인지**: 사용자가 특별히 값을 지정하지 않았을 때 어떤 값이 기본으로 사용되는지 알 수 있습니다.

`GridSearchCV`와 같은 도구를 사용하여 최적의 하이퍼파라미터를 찾을 때, 이 `get_params()`를 통해 확인한 파라미터 이름을 `param_grid`에 사용하게 됩니다.

In [ ]:
LogisticRegression().get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'deprecated',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [ ]:
all_models(X_train, y_train).T

/usr/local/lib/python3.11/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


0   \
mean_fit_time                                               0.006926   
std_fit_time                                                 0.00277   
mean_score_time                                             0.002828   
std_score_time                                               0.00102   
param_C                                                         0.01   
param_penalty                                                     l1   
param_solver                                               liblinear   
params             {'C': 0.01, 'penalty': 'l1', 'solver': 'liblin...   
split0_test_score                                           0.642857   
split1_test_score                                           0.642857   
split2_test_score                                           0.642857   
split3_test_score                                           0.769231   
split4_test_score                                           0.692308   
split5_test_score                                           0.692308   
split6_test_score                                           0.615385   
split7_test_score                                           0.692308   
split8_test_score                                           0.615385   
split9_test_score                                           0.538462   
mean_test_score                                             0.654396   
std_test_score                                              0.058637   
rank_test_score                                                  8.0   
mean_fit_time                                               0.002622   
std_fit_time                                                0.001561   
mean_score_time                                             0.002957   
std_score_time                                              0.000288   
param_metric                                               euclidean   
param_n_neighbors                                                  3   
param_weights                                                uniform   
params             {'metric': 'euclidean', 'n_neighbors': 3, 'wei...   
split0_test_score                                           0.714286   
split1_test_score                                           0.857143   
split2_test_score                                           0.642857   
split3_test_score                                           0.692308   
split4_test_score                                           0.846154   
split5_test_score                                           0.692308   
split6_test_score                                           0.692308   
split7_test_score                                           0.846154   
split8_test_score                                           0.769231   
split9_test_score                                           0.846154   
mean_test_score                                              0.75989   
std_test_score                                              0.078325   
rank_test_score                                                   11   

                                                                  1   \
mean_fit_time                                               0.005739   
std_fit_time                                                0.001762   
mean_score_time                                             0.002962   
std_score_time                                              0.001252   
param_C                                                         0.01   
param_penalty                                                     l2   
param_solver                                               liblinear   
params             {'C': 0.01, 'penalty': 'l2', 'solver': 'liblin...   
split0_test_score                                                1.0   
split1_test_score                                           0.857143   
split2_test_score                                           0.857143   
split3_test_score                                                1.0   
split4_test_score                

### GridSearchCV 2
-📦 모델의 성능을 높이기 위해,
- 우리가 정해줘야 하는 숫자들(=하이퍼파라미터)을
- 자동으로 여러 번 시험해보고,
- 제일 잘 되는 조합을 골라주는 도구예요.

In [ ]:
# param_grid는 dict 또는 dictionary로 된 list를 인자로 받을 수 있다.
grid = GridSearchCV(KNeighborsClassifier(), [{'n_neighbors':[2,3,4,5],
                                              'weights':['uniform', 'distance']}])

In [ ]:
grid.fit(wine.iloc[:, :-1], wine.target)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid=[{'n_neighbors': [2, 3, 4, 5],
                          'weights': ['uniform', 'distance']}])

In [ ]:
# params를 보면 매번 실행할때마다 사용되는 하이퍼 파라미터을 기록한다.
# 위에서 param_grid를 리스트 안에 딕셔너리 형태로 넣었다.
# 딕셔너리 안에서 조합이 구해진다.
# 다른 딕셔너리에 들어간 하이퍼파라미터와는 조합을 만들지 않는다.
pd.DataFrame(grid.cv_results_).T

0  \
mean_fit_time                                       0.00929   
std_fit_time                                          0.004   
mean_score_time                                    0.019883   
std_score_time                                     0.007518   
param_n_neighbors                                         2   
param_weights                                       uniform   
params             {'n_neighbors': 2, 'weights': 'uniform'}   
split0_test_score                                  0.611111   
split1_test_score                                  0.611111   
split2_test_score                                  0.638889   
split3_test_score                                  0.657143   
split4_test_score                                       0.8   
mean_test_score                                    0.663651   
std_test_score                                     0.070381   
rank_test_score                                           8   

                                                           1  \
mean_fit_time                                       0.010195   
std_fit_time                                        0.003782   
mean_score_time                                     0.008896   
std_score_time                                      0.003514   
param_n_neighbors                                          2   
param_weights                                       distance   
params             {'n_neighbors': 2, 'weights': 'distance'}   
split0_test_score                                   0.805556   
split1_test_score                                   0.638889   
split2_test_score                                   0.666667   
split3_test_score                                   0.685714   
split4_test_score                                   0.828571   
mean_test_score                                     0.725079   
std_test_score                                      0.076912   
rank_test_score                                            1   

                                                          2  \
mean_fit_time                                      0.006419   
std_fit_time                                       0.003829   
mean_score_time                                    0.008276   
std_score_time                                     0.003474   
param_n_neighbors                                         3   
param_weights                                       uniform   
params             {'n_neighbors': 3, 'weights': 'uniform'}   
split0_test_score                                  0.638889   
split1_test_score                                  0.694444   
split2_test_score                                  0.666667   
split3_test_score                                  0.657143   
split4_test_score                                  0.857143   
mean_test_score                                    0.702857   
std_test_score                                     0.079204   
rank_test_score                                           5   

                                                           3  \
mean_fit_time                                       0.008471   
std_fit_time                                        0.002497   
mean_score_time                                     0.006319   
std_score_time                                      0.002987   
param_n_neighbors                                          3   
param_weights                                       distance   
params             {'n_neighbors': 3, 'weights': 'distance'}   
split0_test_score                                   0.722222   
split1_test_score                                   0.722222   
split2_test_score                                   0.666667   
split3_test_score                                   0.657143   
split4_test_score                                   0.857143   
mean_test_score                                     0.725079   
std_test_score                                      0.071393   
rank_test_score                                        

## ③ GridSearchCV + Pipeline
(Pipeline은 전체 머신러닝 작업 흐름(전처리 → 모델 학습 등)을 한 덩어리로 묶어주는 도구)

## <GridSearchCV와 Pipeline을 연동하기>
- Pipeline과 GridSearchCV를 연동하면,
- 전처리부터 모델 학습까지의 전체 과정을
- 자동화 + 튜닝 + 교차검증까지 한 번에 처리할 수 있기 때문이에요.

### 쉽게설명
- 예를 들어, 여러분이 꽃의 품종을 분류하는 모델을 만든다고 해요.
- 그런데 데이터에는 스케일 차이가 큰 특성들이 섞여 있어요.
- 그래서 StandardScaler()로 스케일링을 먼저 하고
- KNN, SVM 같은 모델을 학습해야 해요.

### 이런 과정이 보통 이렇게 나뉘죠:

- 1. 데이터 스케일링 (StandardScaler)

- 2. 모델 선택 및 학습 (KNN)

- 3. 하이퍼파라미터 튜닝 (GridSearchCV)

- 4.  교차검증

### 그런데 문제는?
- 하이퍼파라미터 튜닝을 위해 GridSearchCV를 쓰면,
데이터를 나눠서 여러 번 훈련하고 평가해야 해요.

- 그런데 그때마다 스케일링도 반드시 훈련데이터에 맞춰 다시 해야 해요!

- 스케일러를 전체 데이터에 fit 한 후 나누면 → 데이터 누수(data leakage) 발생!

-  그래서 Pipeline이 필요!

> * 데이터 누수란?,
- AI가 시험(테스트 데이터)을 보기 전에 미리 답(정보)을 슬쩍 본 상태를 말해요!**
- *그래서 공정한 평가를 위해, 기준(fit)은 훈련 데이터로만!**



### ✏️ `fit()` = 공부/훈련 (AI가 훈련 데이터로 공부함)

- 이건 **훈련 데이터를 보면서**,
- **평균이 얼마고, 얼마나 퍼져 있는지를 기억하는 과정**이에요.

예:

“나는 지난번 수학 시험에서 평균 60점이었어. 다음엔 70점 맞아야지!”

---

### 📝 `transform()` = 테스트

- 이제 ****시험 문제(테스트 데이터)****가 나왔어요.
- 공부한 내용을 바탕으로 **시험을 공정하게 봐야 해요.**
- 시험지는 새롭지만, **공부할 때 배운 기준(평균 등)**만 적용해서 푸는 거예요.

---

## ❗ 데이터 누수 = 시험 전에 시험 문제를 미리 본 것!

만약 우리가 **시험지를 미리 보고 공부 기준을 세웠다면?**

- 시험지를 보고 “아~ 이건 자주 나오는 문제네!” 하면서
- **공부 기준 자체를 바꿔버렸다면?**
- 그건 **공정한 시험이 아니죠!!** → 즉, **컨닝이에요!**


In [ ]:
from sklearn.pipeline import Pipeline

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
KNeighborsClassifier().get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [ ]:
pipe = Pipeline([('ss', StandardScaler()), ('knn', KNeighborsClassifier())])

In [ ]:
grid = GridSearchCV(pipe, {'knn__n_neighbors':[2,3,4,5]})

In [ ]:
# GridSearchCV 또는 유사한 코드에서 Pipeline 객체의 하위 단계(estimator)의 파라미터를 잘못 지정했기 때문이다.
# ValueError
# Check the list of available parameters with `estimator.get_params().keys()`
grid.fit(wine.iloc[:,:-1], wine.target)

GridSearchCV(estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('knn', KNeighborsClassifier())]),
             param_grid={'knn__n_neighbors': [2, 3, 4, 5]})

In [ ]:
# pipe의 내부를 살펴보자.
vars(pipe)

{'steps': [('ss', StandardScaler()), ('knn', KNeighborsClassifier())],
 'transform_input': None,
 'memory': None,
 'verbose': False}

In [ ]:
# pipe에서 knn 알고리즘만 꺼내서 get_params를 확인해본다.
# n_neighbors가 맞다.
pipe.steps[-1][1].get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [ ]:
# pipe 자체의 get_params를 확인해보자.
# Pipeline은 알고리즘의 하이퍼 파라미터 이름을 맹글링한다. (Pipeline의 docstring 참조)
pipe.get_params()


In [ ]:
# mangling된 이름으로 하이퍼 파라미터 이름을 바꾸었다.
grid = GridSearchCV(pipe, {'knn__n_neighbors':[2,3,4,5]})

In [ ]:
# Pipeline을 거쳐서 GridSeaerchCV까지 한번에 사용할 수 있다.
grid.fit(wine.iloc[:, :-1], wine.target)

GridSearchCV(estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('knn', KNeighborsClassifier())]),
             param_grid={'knn__n_neighbors': [2, 3, 4, 5]})

### GridSearchCV로 최적의 알고리즘 찾기

`GridSearchCV`는 최적의 하이퍼 파라미터는 찾을 수 있지만, 최적의 알고리즘을 찾는 기능은 공식적으로 제공하지 않는다.  
그러나, scikit-learn의 고수들이 알고리즘도 찾아주는 꼼수를 발견하였다.

In [ ]:
pipe = Pipeline([('ss', StandardScaler()), ('clf', KNeighborsClassifier())])

In [ ]:
# pipe에서 clf를 하이퍼 파라미터로 취급하여, 알고리즘을 바꾼다.
grid = GridSearchCV(pipe, {'clf':[KNeighborsClassifier(), LogisticRegression()]})

In [ ]:
grid.fit(wine.iloc[:, :-1], wine.target)

GridSearchCV(estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('clf', KNeighborsClassifier())]),
             param_grid={'clf': [KNeighborsClassifier(), LogisticRegression()]})

In [ ]:
# 알고리즘을 바꿔가면서 GridSearchCV하게 만들 수 있다.
pd.DataFrame(grid.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.004261      0.001152         0.003483        0.000197   
1       0.013304      0.008490         0.002770        0.000124   

                param_clf                           params  split0_test_score  \
0  KNeighborsClassifier()  {'clf': KNeighborsClassifier()}           0.944444   
1    LogisticRegression()    {'clf': LogisticRegression()}           0.972222   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.944444           0.972222           1.000000           0.885714   
1           0.972222           1.000000           0.971429           1.000000   

   mean_test_score  std_test_score  rank_test_score  
0         0.949365        0.037911                2  
1         0.983175        0.013741                1

### 알고리즘마다 하이퍼 파라미터 찾기

In [ ]:
# param_grid를 list 안의 dict로 넣으면, dictionary끼리만 조합을 만든다.
# 최적의 알고리즘과 최적의 하이퍼 파라미터를 동시에 찾을 수 있다.
grid = GridSearchCV(pipe, [{'clf':[KNeighborsClassifier()],
                            'clf__n_neighbors':[2,3,4,5]},
                           {'clf':[LogisticRegression()]}])

In [ ]:
grid.fit(wine.iloc[:, :-1], wine.target)

GridSearchCV(estimator=Pipeline(steps=[('ss', StandardScaler()),
                                       ('clf', KNeighborsClassifier())]),
             param_grid={'clf': [KNeighborsClassifier(), LogisticRegression()]})

In [ ]:
# knn의 하이퍼 파라미터는 knn에서만 사용되었다.
pd.DataFrame(grid.cv_results_).T

0  \
mean_fit_time                             0.004275   
std_fit_time                              0.000968   
mean_score_time                           0.003739   
std_score_time                            0.000444   
param_clf                   KNeighborsClassifier()   
params             {'clf': KNeighborsClassifier()}   
split0_test_score                         0.944444   
split1_test_score                         0.944444   
split2_test_score                         0.972222   
split3_test_score                              1.0   
split4_test_score                         0.885714   
mean_test_score                           0.949365   
std_test_score                            0.037911   
rank_test_score                                  2   

                                               1  
mean_fit_time                           0.009067  
std_fit_time                            0.000898  
mean_score_time                         0.002803  
std_score_time                           0.00003  
param_clf                   LogisticRegression()  
params             {'clf': LogisticRegression()}  
split0_test_score                       0.972222  
split1_test_score                       0.972222  
split2_test_score                            1.0  
split3_test_score                       0.971429  
split4_test_score                            1.0  
mean_test_score                         0.983175  
std_test_score                          0.013741  
rank_test_score                                1